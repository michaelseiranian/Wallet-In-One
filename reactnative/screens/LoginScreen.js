import { StatusBar } from "expo-status-bar";
import {
  StyleSheet,
  Text,
  View,
  ScrollView,
  Button,
  TextInput,
  Alert,
} from "react-native";
import * as Notifications from "expo-notifications";
import { useTheme } from 'reactnative/src/theme/ThemeProvider';
import {styles} from 'reactnative/screens/All_Styles.style.js';

import { useContext, useState } from "react";
import { userContext } from "../data";

import { login } from "../authentication";


import * as SecureStore from "expo-secure-store";

export default function SignUpScreen({ navigation }) {

  const {dark, colors, setScheme} = useTheme();

  const [user, setUser] = useContext(userContext);

  const [username, setUsername] = useState();
  const [password, setPassword] = useState();

  const [errors, setErrors] = useState({});

  const loginHandler = async () => {
    var response = await login(username, password, user, setUser);
    if (response.status == "Error") {
      Alert.alert("Error", response.body["message"]);
    } else if (response.status == 400 && response.body) {
      setErrors(response.body);

      if (response.body["non_field_errors"]) {
        Alert.alert("Error", "Login Error");
      }
    } else {
      console.log("Login Successful")
      await sendLogInNotification();
    }
  };
Notifications.setNotificationHandler({
    handleNotification: async () => {
      return {
        shouldShowAlert: true,
        shouldPlaySound: true,
        shouldSetBadge: true,
      };
    },
  });

const sendLogInNotification = async () => {
    const notificationEnabled = await SecureStore.getItemAsync(
      "notificationSettings"
    );
    console.log("notificationSettings value:", notificationEnabled)
    if (notificationEnabled === "true") {
      await Notifications.requestPermissionsAsync();
      await Notifications.scheduleNotificationAsync({
        content: {
          title: "You have successfully logged in!",
          body: "You can now access all the features of the app.",
        },
        trigger: null,
      });
    }

  };

  const inputStyle = (name) => {
    if (name in errors || "non_field_errors" in errors) {
      return [stylesInternal.input, stylesInternal.error];
    }
    return [stylesInternal.input];
  };

  function ErrorMessage(props) {
    if (props.name in errors) {
      return (
        <>
          {errors[props.name].map((value, index) => {
            return (
              <Text key={index} style={stylesInternal.errorText}>
                {value}
              </Text>
            );
          })}
        </>
      );
    }
    return null;
  }

  return (
    <ScrollView style={[stylesInternal.container, styles(dark, colors).container]}>
      <StatusBar style="auto" />

      <Text style={[stylesInternal.text, styles(dark, colors)?.text]}>Username:</Text>
      <TextInput
        autoCapitalize='none'
        autoCorrect={false}
        style={[inputStyle("username"), styles(dark, colors)?.input, {color: colors?.text}]}
        onChangeText={setUsername}
        testID="username"
      />
      <ErrorMessage name="username"></ErrorMessage>

      <Text style={[stylesInternal.text, styles(dark, colors)?.text]}>Password:</Text>
      <TextInput
        autoCapitalize='none'
        autoCorrect={false}
        style={[inputStyle("password"), styles(dark, colors)?.input, {color: colors?.text}]}
        onChangeText={setPassword}
        secureTextEntry={true}
        testID="password"
      />
      <ErrorMessage name="password"></ErrorMessage>
      <ErrorMessage name="non_field_errors"></ErrorMessage>

      <View style={stylesInternal.parent}>
        <Button style={stylesInternal.button} title="Log In" onPress={loginHandler} />
      </View>
    </ScrollView>
  );
}

const stylesInternal = StyleSheet.create({
  container: {
    width: "100%",
    padding: 30,
  },
  parent: {
    marginTop: 20,
    flex: 1,
    width: "100%",
    alignSelf: "flex-start",
  },
  text: {
    marginTop: 10,
    marginBottom: 10,
  },
  input: {
    height: 40,
    width: "100%",
    borderWidth: 0.5,
    padding: 10,
    borderColor: "gray",
    borderRadius: 5,
  },
  button: {},
  error: {
    borderColor: "red",
  },
  errorText: {
    marginLeft: 10,
    color: "red",
  },
});