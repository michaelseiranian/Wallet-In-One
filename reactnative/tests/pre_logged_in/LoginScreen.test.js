import React from 'react';
import { render, screen, fireEvent, act, within} from '@testing-library/react-native';
import LoginScreen from '../../screens/LoginScreen';


jest.mock('react', () => {
  const ActualReact = jest.requireActual('react')
  return {
    ...ActualReact,
    useContext: () => (ActualReact.useState({'signedIn': false})),
  }
})

var empty = {
  "username": [
      "This field is required."
  ],
  "password": [
      "This field is required."
  ],
}

var invalid = {
  "non_field_errors": [
      "Unable to log in with provided credentials."
  ]
}

var success = {
  "token": "123"
}

jest.mock('../../authentication', () => {
  return {
    login: jest.fn((username, password) => {
      if (!username && !password){
        return Promise.resolve({status:400,body: empty})
      }
      if (username == "john doe" && password == "password123"){
        return Promise.resolve({status:200,body: success})
      }
      else{
        return Promise.resolve({status:400,body: invalid})
      }
    })
}})

describe('<LoginScreen />', () => {
  it('nothing entered', async () => {

    var renderer = render(<LoginScreen />)

    await act( async () => {
      fireEvent.press(await screen.getByText('Log In'));
    });

    const errors = await screen.findAllByText("This field is required.");
    expect(errors.length).toBe(2);
  });

  it('wrong credentials ', async () => {

    var renderer = render(<LoginScreen />)

    await act( async () => {
      fireEvent.changeText(await screen.getByTestId('username'), 'wrongusername');
      fireEvent.changeText(await screen.getByTestId('password'), 'wrongpassword');

      fireEvent.press(await screen.getByText('Log In'));
    });
    

    const errors = await screen.findAllByText("Unable to log in with provided credentials.");
    expect(errors.length).toBe(1);
  });

  it('successful login ', async () => {

    var renderer = render(<LoginScreen />)

    await act( async () => {
      fireEvent.changeText(await screen.getByTestId('username'), 'john doe');
      fireEvent.changeText(await screen.getByTestId('password'), 'password123');

      fireEvent.press(await screen.getByText('Log In'));
    });
    
    const errors1 = await screen.queryAllByText("Unable to log in with provided credentials.");
    const errors2 = await screen.queryAllByText("This field is required.");

    expect(errors1).toEqual([])
    expect(errors2).toEqual([])
  });

  it('log in snapshot test', () => {
    const renderer = render(<LoginScreen />)
    expect(renderer).toMatchSnapshot();
  });
});
