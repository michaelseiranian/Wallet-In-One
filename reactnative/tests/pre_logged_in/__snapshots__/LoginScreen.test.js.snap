// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginScreen /> log in snapshot test 1`] = `
<RCTScrollView
  style={
    [
      {
        "padding": 30,
        "width": "100%",
      },
      {
        "backgroundColor": undefined,
        "flex": 1,
      },
    ]
  }
>
  <View>
    <Text
      style={
        [
          {
            "marginBottom": 10,
            "marginTop": 10,
          },
          {
            "color": undefined,
          },
        ]
      }
    >
      Username:
    </Text>
    <TextInput
      autoCapitalize="none"
      autoCorrect={false}
      onChangeText={[Function]}
      style={
        [
          [
            {
              "borderColor": "gray",
              "borderRadius": 5,
              "borderWidth": 0.5,
              "height": 40,
              "padding": 10,
              "width": "100%",
            },
          ],
          {
            "borderColor": "gray",
            "borderRadius": 5,
            "borderWidth": 0.5,
            "height": 40,
            "marginBottom": 5,
            "marginTop": 5,
            "padding": 10,
            "width": "100%",
          },
          {
            "color": undefined,
          },
        ]
      }
      testID="username"
    />
    <Text
      style={
        [
          {
            "marginBottom": 10,
            "marginTop": 10,
          },
          {
            "color": undefined,
          },
        ]
      }
    >
      Password:
    </Text>
    <TextInput
      autoCapitalize="none"
      autoCorrect={false}
      onChangeText={[Function]}
      secureTextEntry={true}
      style={
        [
          [
            {
              "borderColor": "gray",
              "borderRadius": 5,
              "borderWidth": 0.5,
              "height": 40,
              "padding": 10,
              "width": "100%",
            },
          ],
          {
            "borderColor": "gray",
            "borderRadius": 5,
            "borderWidth": 0.5,
            "height": 40,
            "marginBottom": 5,
            "marginTop": 5,
            "padding": 10,
            "width": "100%",
          },
          {
            "color": undefined,
          },
        ]
      }
      testID="password"
    />
    <View
      style={
        {
          "alignSelf": "flex-start",
          "flex": 1,
          "marginTop": 20,
          "width": "100%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            [
              {},
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Log In
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
